snuba/admin/clickhouse/migration_checks.py

```diff
@@ -0,0 +1,215 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import List, Sequence, Tuple
+
+from snuba.admin.migrations_policies import get_migration_group_polices
+from snuba.migrations.groups import MigrationGroup
+from snuba.migrations.policies import MigrationPolicy
+from snuba.migrations.runner import MigrationDetails, MigrationKey
+from snuba.migrations.status import Status
+
+
+class Reason(Enum):
+    NO_REASON_NEEDED = ""
+    ALREADY_RUN = "already run"
+    NOT_RUN_YET = "can't reverse if not already run"
+    NEEDS_SUBSEQUENT_MIGRATIONS = "subsequent migrations must be reversed first"
+    NEEDS_EARLIER_MIGRATIONS = "earlier migrations must run first"
```

Comment:
This class is mixing together reasons for running and reversing migrations. It feels like the data model is not representing your states in the right way. You could have a lot of inconsistent states. Like `can_run` returning an instance of `ResultReason.NOT_RUN_YET`.

Since you have two methods: `can_run` and `can_reverse`, You can have two different result types. The fields in the enum has only one value that applies to both.

Reply:
updated to separate out result types and reason types